### PR TITLE
Module 3

### PR DESCRIPTION
**Pull request by anniedroid**
1) Main.c
	a) Load all data - working
	b)Interface added for BOOKS pillar
		* View book - working
		* Search book - working
		* filter book - not working, codes got skipped, not prompting for input 
		* delete book: working but the returned ints that count numbers of deleted books/ copies are not accurate, returning large/negative int
	c) USERS/TRANS pillars pending
2) Auth.c/h
	change login function to return struct User* instead of bool
3) Books.c/h
	a) Added missing filter functions for counts/likes/etc(we didn't do them before because they're the same logic)
	b) Adding 2 delete functions for books/book copies
4) Users.h
	a) Adding fillterByFullName function to header file to use in main (to get users ID from user name for filterBook function)
              *

### CODE REVIEW

File: books.c [hunk 1]
```diff
@@ -277,5 +333,62 @@ void freeBookList(Book* head)
...
+
+    while (current) {
+        if (isMatch(current->ISBN, input_ISBN)) {
+            if (!previous) { // if match @ head
+                library = current->next;
+            } else {
+                previous->next = current->next;
+            }
+            free(current);
+            return 1; // Book found and deleted, return because there should only 1 book match
```
Comment (saadali18) [inline books.c]: As discussed in a live class, we are not going to delete record permanently from our system, we will prefer soft delete (status: inactive/deleted)

File: main.c [hunk 2]
```diff
@@ -9,128 +9,167 @@
...
+
+    printf("Welcome to the Library Management System!\n");
+    char* login_name;
+    char* password;
+    User* current_user;
+    do {
+        login_name = getLoginName();
+        password = getPassword();
+        current_user = login(login_name, password);
+    } while (!current_user);
```
Comment (saadali18) [inline main.c]: This is an infinite loop, add some exit point to it.

**Review by saadali18 — CHANGES_REQUESTED** [review]

**Comment by anniedroid** [comment]
Librarian interface: book cases code running, bug fixed.

File: main.c [hunk 3]
```diff
@@ -9,128 +10,413 @@
...
-    printRecord(transaction_record_head);
     loadTransactions(transaction_record_head, &transaction_list);
-    printTransactions(transaction_list);
-    saveTransactions(transaction_file, transaction_list);
 
+    printf("Welcome to the Library Management System!\n");
+    // AUTHENTICATION
+    char* log_in_name;
+    char* password;
+    int attempt = 0;
```
Comment (saadali18) [inline main.c:30]: Although it will work in our use case, but practically we need to implement it other way

File: main.c [hunk 4]
```diff
@@ -9,128 +10,413 @@
...
+        current_user = login(log_in_name, password);
+        attempt++;
+    } while (!current_user && attempt < 3);
+
+    if (attempt == 3) {
+        printf("Multiple failed login attempts detected. Please contact admin\n");
+        exit(0);
+    }
+    // LIBRARIAN
+    if (current_user->title == librarian) {
```
Comment (saadali18) [inline main.c:44]: current_user -> title ? or current_user -> credential
kindly confirm

File: main.c [hunk 5]
```diff
@@ -9,128 +10,413 @@
...
+                        else    printf("No book found!\n");
+                        // Delete all copies of source book
+                        int deleted_book_copies = deleteBookCopies(ISBN_to_delete);
+                        if (deleted_book_copies > 0)    printf("%i book copies deleted\n", deleted_book_copies);
+                        else    printf("No book copies found!\n");
+                        free(ISBN_to_delete);
+                        break;
+                    default:
+                        break;
+                }
```
Comment (saadali18) [inline main.c:168]: I think we need to remove/inactive the (current) transactions of deleted book as well

File: main.c [hunk 6]
```diff
@@ -9,128 +10,413 @@
...
+                        break;
+                    case 6:
+                        printf("Enter user ID to be deleted: ");
+                        int id = 0;
+                        scanf("%i", &id);
+                        int deleted_user = deleteUser(id);
+                        if (deleted_user > 0)  printf("%i user deleted!\n", deleted_user);
+                        else    printf("No user found!\n");
+                        break;
+                }
```
Comment (saadali18) [inline main.c:305]: If user is author, his books should be inactive as well 
if user is customer, all his transactions should be inactive and book copies count should update

File: main.c [hunk 7]
```diff
@@ -9,128 +10,413 @@
...
+                break;
+        } // END OF SWITCH CATEGORY
+    } // END OF LIBRARIAN
+
+    /*// SAVE TO FILE BEFORE EXIT
+    saveBooks(book_file, library);
+    saveBookCopies(book_copy_file, inventory);
+    saveUsers(user_file, user_list_head);
+    saveTransactions(transaction_file, transaction_list);*/
+
```
Comment (saadali18) [inline main.c]: why this code is commented?

**Review by saadali18 — CHANGES_REQUESTED** [review]

**Pull request merged**